### DISCUSSION TITLE
Kiali Operator not showing Jaeger Traces

### DISCUSSION THREAD

**Discussion by AnthonyGurg** [kiali/kiali#3566] [Q&A]
Hello there, 

I've been using the Kiali and Jaeger operators on OCP and I've been struggling to get the traces showing on Kiali.

I'm able to view the traces on Jaeger without issue by going to the Jaeger route directly but not on Kiali. On Kiali I see this error appear: `Error with internal connection with Jaeger` with no further details in the pod logs or on Kiali.

The default in_cluster_url was  `https://jaeger-query.istio-system.svc` but I had created my own jaeger instance "simple-prod2".

My Kiali config has the following tracing configuration:

```
tracing:
  auth:
    password: >- 
      password
    type: basic
    use_kiali_token: false
    username: internal
  enabled: true
  in_cluster_url: 'https://simple-prod2-query.istio-system.svc'
  namespace: istio-system
  service: ''
  url: 'https://simple-prod2-istio-system.apps.kiali.domain.com'
```

I'd appreciate some help troubleshooting this issue. Thanks.

**Comment by jotak** [comment]
Hi @AnthonyGurg ,
The `in_cluster_url` that Kiali expects is the URL that should be accessible from the Kiali pod. You can try `curl` this URL from the pod to make sure it is correctly configured.

- First, check that the jaeger host is correct:

```
kubectl exec <kiali pod> -n istio-system -- curl https://simple-prod2-query.istio-system.svc
```

(This URL suggests that `simple-prod2-query` is your Jaeger service name, it's deployed in namespace `istio-system`, and it listens on https / port 443 ; if incorrect, this should be adapted. For instance with istio 1.8 demo deployment, the jaeger query url is: `http://tracing.istio-system.svc/jaeger` (http / port 80) )

If the URL is correct, you should see some html and javascript as a response:

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  2761    0  2761    0     0   552k      0 --:--:-- --:--:-- --:--:--  674k
<!doctype html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width,initial-scale=1"><meta http-equiv="cache-control" content="max-age=0"/><meta http-equiv="cache-control" content="no-cache"/><meta http-equiv="expires" content="0"/><meta http-equiv="expires" content="Tue, 01 Jan 1980
(etc.)
```

- Secondly, to make sure Kiali hits the correct path endpoint in jaeger query, you can `curl` jaeger's API and make sure json is returned, by appending `/api/traces` to the URL.

E.g., my Jaeger URL being `http://tracing.istio-system.svc/jaeger` :

```
kubectl exec kiali-7d5cb68b45-sq2dc -n istio-system -- curl http://tracing.istio-system.svc/jaeger/api/traces
```
produces:
```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   109  100   109    0     0  22222      0 --:--:-- --:--:-- --:--:-- 27250
{"data":null,"total":0,"limit":0,"offset":0,"errors":[{"code":400,"msg":"parameter 'service' is required"}]}
```
whereas a wrong path would in general still produce html (e.g. if omitting the `/jaeger` part)

**Reply by AnthonyGurg** [reply]
Hi @jotak, thanks for the response. So I curled both URL's and received:

> curl performs SSL certificate verification by default, using a "bundle"
>  of Certificate Authority (CA) public keys (CA certs). If the default
>  bundle file isn't adequate, you can specify an alternate file
>  using the --cacert option.
> If this HTTPS server uses a certificate signed by a CA represented in
>  the bundle, the certificate verification probably failed due to a
>  problem with the certificate (it might be expired, or the name might
>  not match the domain name in the URL).
> If you'd like to turn off curl's verification of the certificate, use
>  the -k (or --insecure) option.
> command terminated with exit code 60

When using --insecure, I was able to get the HTML/JS for both files, the second URL was showing HTML/JS for the login page so I guess I need to properly configure the login so I can receive the JSON. For testing purposes, is there a way I can get Kiali to ignore the SSL issues? 

Thanks.

**Reply by jotak** [reply]
Yes, the config is described here: https://github.com/kiali/kiali-operator/blob/81417333ea9cbfe600f48ddc85d57f99e5e3f545/deploy/kiali/kiali_cr.yaml#L547-L586
You can use `insecure_skip_verify` to allow an insecure connection.

**Reply by AnthonyGurg** [reply]
Sorry I forgot to mention that I did try insecure_skip_verify before my last question. I was still getting the same errors so I went back to the original Jaeger instance "Jaeger". It's possible my credentials for my simple-prod Jaeger instance might have been wrong and also, it seems that Kiali kept reverting back to the "jaeger" instance for the URL and was setting a password by default.

The "Error with internal connection" doesn't appear when I revert back to the Jaeger instance but traces still do not appear on Kiali while appearing on the external Jaeger URL. My current configuration:

```
    tracing:
      auth:
        ca_file: ''
        insecure_skip_verify: true
        password: >-
          <password being set automatically>
        token: ''
        type: basic
        use_kiali_token: false
        username: internal
      enabled: true
      in_cluster_url: 'https://jaeger-query.istio-system.svc'
      namespace: istio-system
      service: ''
      url: 'https://jaeger-istio-system.apps.kiali.domain.com'
      whitelist_istio_system:
        - jaeger-query
        - istio-ingressgateway
```

Would you have any other ideas? Thanks.

**Reply by jotak** [reply]
Which version of Kiali do you have? The configuration suggests it's a pretty old one (fields like `namespace` and `service` in tracing have been removed for a while), perhaps upgrading to a more recent version could help?
Also, Kiali keeping reverting config would suggest you are modifying the configmap instead of the Kiali CR ? When installed with the operator, the configmap shouldn't be manually modified, as it's generated from CR. The configuration structure is pretty much the same though.

**Reply by AnthonyGurg** [reply]
I'm using the Kiali Operator provided by Redhat version: 1.24.3. I can try switching over to the Community Kiali Operator and see if anything changes.

And I was only updating the CR never the configmap. I'm not sure if the errors were triggering the changes but the password field was being reset and the URL field would be changed from simple-prod2-istio to jaeger-istio...

**Reply by jotak** [reply]
I'm going to request help from @jmazzitelli here because I don't know how the values you set in CR could be overwritten.
Also, the old config fields that I mentioned pre-date 1.24.x by far (like one year before), so, perhaps there was an upgrade at some point that messed up a couple of things? Can you try reinstalling the operator (v1.24.3 should be fine wrt jaeger interactions, no need to switch to community)

**Reply by AnthonyGurg** [reply]
The CR being overwritten wasn't really an issue for me as it was only happening for the URL and not the in_cluster_url.

However, with a fresh cluster and installing the operator provided by Redhat, the issue remains and the CR that's automatically generated is:

```    tracing:
      auth:
        password: >-
          <PASSWORD>
        type: basic
        use_kiali_token: false
        username: internal
      enabled: true
      in_cluster_url: 'https://jaeger-query.istio-system.svc'
      namespace: istio-system
      service: ''
      url: 'https://jaeger-istio-system.apps.domain.com'
```

And that was before I made any changes, all I did was install Jaeger, Kiali, Elasticsearch and the ServiceMesh(followed this [guide](https://docs.openshift.com/container-platform/4.4/service_mesh/v1x/installing-ossm.html)). So it does still include "namespace", which is also shown at: https://github.com/kiali/kiali-operator/blob/v1.24/deploy/kiali/kiali_cr.yaml

**Reply by FilipB** [reply]
> And that was before I made any changes, all I did was install Jaeger, Kiali, Elasticsearch and the ServiceMesh(followed this [guide](https://docs.openshift.com/container-platform/4.4/service_mesh/v1x/installing-ossm.html)). So it does still include "namespace", which is also shown at: https://github.com/kiali/kiali-operator/blob/v1.24/deploy/kiali/kiali_cr.yaml

I will try to reproduce on OCP 4.4, could you please add your SMCP CR?
It's the step 7. in the install guide you referenced - 
"On the Create Service Mesh Control Plane page, modify the YAML for the default ServiceMeshControlPlane template as needed."

**Reply by jotak** [reply]
@AnthonyGurg I guess you're using OCP 4.4 and ServiceMesh 1.x, is that correct? That would explain some of the config fields that aren't expected, and that's regardless of the version of the Kiali operator (because the Istio operator infers some values into the Kiali CR, like - you guess it - jaeger URL and password ; but you can change them in the ServiceMeshControlPlane resource as @FilipB says).

@jmazzitelli can you tell if the operator v1.24.3 supports working with ServiceMesh 1.x?

Also, perhaps in spite of these configuration glitches we can manage having the traces showing up in Kiali, but I'd need to get some additional info.
- First, do you have any suspicious logs in the kiali pod?
- Are traces not showing up for any service?
- Do you get the same for any menu: Applications, Workloads and Services?
- Do you have `app` labels set on pods, as per https://istio.io/latest/docs/ops/deployment/requirements/#pod-requirements ? (the generated traces make use of this "app" label to determine the service name that's stored by jaeger).
Perhaps a screenshot could also help to spot issues if it's something different.

**Reply by AnthonyGurg** [reply]
@FilipB I used the default template when creating the SMCP CR on OCP 4.3 which is:
```
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  version: v2.0
  tracing:
    type: Jaeger
    sampling: 10000
  policy:
    type: Istiod
  telemetry:
    type: Istiod
  addons:
    jaeger:
      install:
        storage:
          type: Memory
    prometheus:
      enabled: true
    kiali:
      enabled: true
    grafana:
      enabled: true
```

@jotak 
I'm on OCP 4.3(I followed the 4.3 guide but the link died - it's the same guide as 4.4) and ServiceMesh 2.0.1. I can go resetup everything on 4.4 though, any version is fine for me if we can get it working there. And the password/url makes sense, thanks.

- I don't have any suspicious logs in the Kiali pod.
- None of the services I've created have any traces showing up on Kiali and it doesn't seem like anything else does either.
- I'm able to view Metrics for the Application
- I do have an app label on my pod

Here is some of the pod config cut down a bit:
```
kind: Pod
apiVersion: v1
metadata:
  generateName: metricsapp-docker7-7bc459f96f-
  annotations:
    openshift.io/scc: anyuid
    prometheus.io/port: '9443'
    openshift.io/generated-by: OpenShiftWebConsole
    k8s.v1.cni.cncf.io/networks: v1-1-istio-cni
    prometheus.io/path: /metrics
    prometheus.io/scheme: https
    sidecar.istio.io/inject: 'true'
    sidecar.istio.io/status: >-
      {"version":"2fc95cd20cf2820cfefbccccc00fa54ad305dd6a6b62ec3ff469ae0b47ad2711","initContainers":null,"containers":["istio-proxy"],"volumes":["istio-envoy","istio-certs"],"imagePullSecrets":null}
    kiali.io/runtimes: myapp
    prometheus.io/scrape: 'true'
  selfLink: >-
    /api/v1/namespaces/jaegertracing/pods/metricsapp-docker7-7bc459f96f-bqmrl
  resourceVersion: '244435462'
  name: metricsapp-docker7-7bc459f96f-bqmrl
  uid: 5156fd2f-5228-445c-8316-60a3a09bc2d4
  creationTimestamp: '2021-01-14T15:40:51Z'
  namespace: jaegertracing
  ownerReferences:
    - apiVersion: apps/v1
      kind: ReplicaSet
      name: metricsapp-docker7-7bc459f96f
      uid: dbbf7f53-3c27-4a3f-83b9-6060efb82b81
      controller: true
      blockOwnerDeletion: true
  labels:
    app: metricsapp-docker7
    pod-template-hash: 7bc459f96f
    security.istio.io/tlsMode: istio
spec:
  restartPolicy: Always
  serviceAccountName: default
  imagePullSecrets:
    - name: default-dockercfg-vwdwp
  priority: 0
  schedulerName: default-scheduler
  enableServiceLinks: true
  terminationGracePeriodSeconds: 30
  nodeName: worker0.kiali.os.domain.com
  securityContext:
    seLinuxOptions:
      level: 's0:c25,c10'
  containers:
    - resources: {}
      terminationMessagePath: /dev/termination-log
      name: metricsapp-docker7
      env:
        - name: JAEGER_REPORTER_LOG_SPANS
          value: 'true'
        - name: JAEGER_SAMPLER_PARAM
          value: '1'
        - name: JAEGER_SAMPLER_TYPE
          value: const
        - name: JAEGER_SERVICE_NAME
          value: metricsapp-docker7.jaegertracing
        - name: JAEGER_PROPAGATION
          value: 'jaeger,b3'
      securityContext:
        capabilities:
          drop:
            - MKNOD
      ports:
        - containerPort: 9080
          protocol: TCP
        - containerPort: 9443
          protocol: TCP
      imagePullPolicy: Always
      volumeMounts:
        - name: default-token-4tnfp
          readOnly: true
          mountPath: /var/run/secrets/kubernetes.io/serviceaccount
```

**Reply by FilipB** [reply]
@AnthonyGurg There are two problems:
- OCP 4.3 is not supported
- SM 2.x is only supported on OCP 4.5 and 4.6

I've just tried to follow your steps on OCP 4.5 and I can see traces for a bookinfo app correctly.
One thing which might be fixed in the documentation is that even when following the install guide for SM 1.x it installs SM 2.x when using default values.

**Reply by AnthonyGurg** [reply]
Thanks for the response Filip, I had no idea that the default CR would install a different version than the guide...

I'll give 4.5 a try and let you know how it goes, thanks.

**Reply by AnthonyGurg** [reply]
Sorry it took me a little while to get everything up and running, I was able to get it to work and show the traces in Kiali. Thanks for the help!

**Comment by jmazzitelli** [comment]
> I'm using the Kiali Operator provided by Redhat version: 1.24.3.

Are you using the OpenShift ServiceMesh 2.0 product? If so, you could go through RedHat support to get "official" support for your issues.

You can't mix-and-match the RedHat operator with upstream server and vice versa.

Red Hat Kiali Operator 1.24.3 is based on the 1.24.x line of Kiali Server so any bleeding-edge features released in later versions are not supported yet. This is most likely your issue. The supported Kiali CR settings for that is here - https://github.com/kiali/kiali-operator/blob/v1.24/deploy/kiali/kiali_cr.yaml

**Reply by AnthonyGurg** [reply]
Sorry my response didn't go through yesterday, I am using the OSSM 2.0 product. I'll open up a ticket if we're unable to get it working, thank you.
